modules/standard/Barrier.chpl

```diff
@@ -0,0 +1,199 @@
+      }
+    }
+
+    /* Block until n tasks have called this method.  If `reusable` is true,
+       reset the barrier to be used again.
+     */
+    inline proc barrier() {
+      on this {
+        const myc = count.fetchSub(1);
+        if myc<=1 {
+          if done.testAndSet() then
+            halt("Too many callers to barrier()");
+          if reusable {
+            count.waitFor(n-1);
+            count.add(1);
+            done.clear();
+          }
+        } else {
+          done.waitFor(true);
+          if reusable {
+            count.add(1);
+            done.waitFor(false);
+          }
+        }
+      }
+    }
+
+    /* Notify the barrier that this task has reached this point. */
+    inline proc notify() {
+      on this {
+        const myc = count.fetchSub(1);
+        if myc<=1 {
+          if done.testAndSet() then
+            halt("Too many callers to notify()");
+        }
+      }
+    }
+
+    /* Wait until `n` tasks have called :proc:`notify`.  If `reusable` is true,
+       reset the barrier to be used again.  Note: if `reusable` is true
+       :proc:`wait` will block until `n` tasks have called it after calling
+       :proc:`notify`.
+     */
+    inline proc wait() {
+      on this {
+        done.waitFor(true);
+        if reusable {
+          const myc = count.fetchAdd(1);
+          if myc == n-1 then
+            done.clear();
+          done.waitFor(false);
+        }
+      }
+    }
+
+    /* return `true` if `n` tasks have called :proc:`notify`
+
+       :rtype: bool
+     */
+    inline proc try() {
+      return done.read();
+    }
+  }
+
+  /* A task barrier implemented using sync and single variables. Can be used
+     as a simple barrier or as a split-phase barrier.
+   */
+  record sBarrier: BarrierBaseType {
+    pragma "no doc"
+    var count: sync int;
+    pragma "no doc"
+    var done: single bool;
+
+    /* Construct a new `n` task Barrier.
+
+       :arg n: The number of tasks that will be involved in the barrier.
+     */
+    proc sBarrier(n: int) {
+      count = n;
+    }
+
+    /* Block until `n` tasks have called this method.
+     */
+    inline proc barrier() {
+      on this {
+        const myc = count;
+        if myc==1 {
+          done = true;
+        } else {
+          if myc < 1 then
+            halt("Too many callers to barrier()");
+          count = myc-1;
+          wait();
+        }
+      }
+    }
+
+    /* Notify the barrier that this task has reached this point. */
+    inline proc notify() {
+      on this {
+        const myc = count;
+        if myc==1 {
+          done = true;
+        } else {
+          if myc < 1 then
+            halt("Too many callers to notify()");
+          count = myc-1;
+        }
+      }
+    }
+
+    /* Wait until `n` tasks have called :proc:`notify`. */
+    inline proc wait() {
+      done;
+    }
+
+    /* return `true` if `n` tasks have called :proc:`notify`
+
+       :rtype: bool
+     */
+    inline proc try() {
```

Comment:
Stylistically, my memory was that we were preferring to declare return types in procedure headers (and perhaps not redundantly use the :rtype: field, similar to how you didn't redundantly mention 'n's type when documenting those arguments.  But I don't feel fully confident that we reached consensus on this, so maybe it's just my personal preference.